ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala

```diff
@@ -809,43 +825,106 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
+          delete from participant_events_create delete_events
           where
             delete_events.event_offset <= $pruneUpToInclusive and
             exists (
               SELECT 1 FROM participant_events_consuming_exercise archive_events
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Create events (only for contracts archived before the specified offset)
-          delete from participant_events_create delete_events
+            )"""
+
+    val retroactiveDivulgencesPruning =
+      if (pruneAllDivulgedContracts)
+        SQL"""
+          -- Retroactive divulgence events
+          delete from participant_events_divulgence delete_events
+          where delete_events.event_offset <= $pruneUpToInclusive"""
+      else
+        SQL"""
+          -- Divulgence events (only for contracts archived before the specified offset)
+          delete from participant_events_divulgence delete_events
           where
             delete_events.event_offset <= $pruneUpToInclusive and
             exists (
               SELECT 1 FROM participant_events_consuming_exercise archive_events
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Exercise events (consuming)
-          delete from participant_events_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-      SQL"""
-          -- Exercise events (non-consuming)
-          delete from participant_events_non_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-    ).foreach(_.execute()(connection))
+            )"""
+
+    {
+      conditionalImmediateDivulgencePruning ++
+        List(
+          retroactiveDivulgencesPruning,
+          createsPruning,
+          consumingExercisesPruning,
+          nonConsumingExercisesPruning,
+        )
+    }.foreach(_.execute()(connection))
+  }
+
+  // TODO divulgence pruning: Check if specialization for Oracle is needed (as it is for H2)
+  protected[CommonStorageBackend] def pruneImmediateDivulgence(
+      pruneUpToInclusive: Offset,
+      pruneAllDivulgedContracts: Boolean,
+  ): List[SimpleSql[Row]] = {
+    import com.daml.platform.store.Conversions.OffsetToStatement
+
+    if (pruneAllDivulgedContracts)
+      List(
+        SQL"""
+          -- Immediate divulgence events
+          WITH
+          	creates_stakeholders AS (
+          		SELECT event_offset, event_sequential_id, unnest(flat_event_witnesses) AS stakeholder
+          		FROM participant_events_create
+          		WHERE event_offset <= $pruneUpToInclusive
+          	),
```

Comment:
This select is missing the condition on the lower-bound of `prunedAllDivulgenceEventsUpToIncluseive` to aid limiting the scan range.

Reply:
Updated according to the discussion had offline.